### PR TITLE
Seaice/njeffery snow rebase

### PR DESCRIPTION
**Pull request by njeffery**
Bringing snow model developments to seaice/develop.

Adds column/ice_snow.F90, additional optional column tracers, and testsuite.snowMods.xml.
Runs BFB with default namelist (snow modifications off).  
Snow module requires a stream input with snow aging properties: "snicar_drdt_bst_fit_60_c070416.nc"

Testing and relations to other Pull Requests should be added as subsequent comments.

See the below examples for more information.
https://github.com/MPAS-Dev/MPAS/pull/930
https://github.com/MPAS-Dev/MPAS/pull/931

### CODE REVIEW

File: src/core_seaice/Registry.xml [hunk 1]
```diff
@@ -92,6 +92,18 @@
 			definition="3"
 			description="Number of spectral intervals in the radiation calculation"
 		/>
+		<dim name="T"
```
Comment (akturner) [inline src/core_seaice/Registry.xml]: Rename

File: src/core_seaice/Registry.xml [hunk 2]
```diff
@@ -92,6 +92,18 @@
...
+			definition="11"
+			name_in_code="nSnowTemperatureIndex"
+		/>
+		<dim name="dTdz"
```
Comment (akturner) [inline src/core_seaice/Registry.xml]: rename

File: src/core_seaice/Registry.xml [hunk 3]
```diff
@@ -92,6 +92,18 @@
...
+			definition="31"
+			name_in_code="nSnowTemperatureGradIndex"
+		/>
+		<dim name="sno_dns"
```
Comment (akturner) [inline src/core_seaice/Registry.xml]: rename

File: src/core_seaice/Registry.xml [hunk 4]
```diff
@@ -639,6 +659,10 @@
 			description="Run the ridging column physics calculation."
 			possible_values="true or false"
 		/>
+		<nml_option name="config_use_column_snow" type="logical" default_value="false" units="unitless"
```
Comment (akturner) [inline src/core_seaice/Registry.xml]: This name is ambiguous. Snow is used whenever column physics is used so what is really being switched on here?

File: src/core_seaice/Registry.xml [hunk 5]
```diff
@@ -677,6 +701,16 @@
 			possible_values="true or false"
 			icepack_name="tr_aero"
 		/>
+		<nml_option name="config_use_snow_density" type="logical" default_value="false" units="unitless"
```
Comment (akturner) [inline src/core_seaice/Registry.xml]: Ambiguous name: Maybe "config_use_effective_snow_density"?

File: src/core_seaice/Registry.xml [hunk 6]
```diff
@@ -1946,6 +2027,17 @@
 			<var name="aerosolSingleScatterAlbedo5band"/>
 		</stream>
 
+		<stream name="SnowAgingPropertiesInput"
+				type="input"
+				filename_template="snicar_drdt_bst_fit_60_c070416.nc"
+				filename_interval="none"
+				input_interval="initial_only"
+				immutable="true">
+			<var name="tau"/>
```
Comment (akturner) [inline src/core_seaice/Registry.xml]: rename

File: src/core_seaice/Registry.xml [hunk 7]
```diff
@@ -1946,6 +2027,17 @@
...
+			<var name="kappa"/>
```
Comment (akturner) [inline src/core_seaice/Registry.xml]: rename

File: src/core_seaice/Registry.xml [hunk 8]
```diff
@@ -1946,6 +2027,17 @@
...
+			<var name="drdsdt0"/>
```
Comment (akturner) [inline src/core_seaice/Registry.xml]: rename

File: src/core_seaice/Registry.xml [hunk 9]
```diff
@@ -2181,6 +2273,8 @@
 		<package name="pkgColumnTracerPonds" description=""/>
 		<package name="pkgColumnTracerLidThickness" description=""/>
 		<package name="pkgColumnTracerAerosols" description=""/>
+		<package name="pkgColumnTracerSnowDensity" description=""/>
```
Comment (akturner) [inline src/core_seaice/Registry.xml]: Add description (I know it seems unfair with nothing else having a description!)

File: src/core_seaice/Registry.xml [hunk 10]
```diff
@@ -2181,6 +2273,8 @@
...
+		<package name="pkgColumnTracerSnowGrainRadius" description=""/>
```
Comment (akturner) [inline src/core_seaice/Registry.xml]: Description

File: src/core_seaice/Registry.xml [hunk 11]
```diff
@@ -2232,6 +2326,30 @@
 		<var name="iceEnthalpy"			type="real"	dimensions="nIceLayers nCategories nCells Time"		name_in_code="iceEnthalpy"/>
 		<var name="iceSalinity"			type="real"	dimensions="nIceLayers nCategories nCells Time"		name_in_code="iceSalinity"/>
 		<var name="snowEnthalpy"			type="real"	dimensions="nSnowLayers nCategories nCells Time"		name_in_code="snowEnthalpy"/>
+		<var name="snowIceMass"			type="real"	dimensions="nSnowLayers nCategories nCells Time"
```
Comment (akturner) [inline src/core_seaice/Registry.xml]: Remove extra spaces between attributes. This is the new style when multiline with descriptions etc.

File: src/core_seaice/Registry.xml [hunk 12]
```diff
@@ -2485,6 +2623,10 @@
 		<var name="iceEnthalpyMask"			type="integer"	dimensions="nIceLayers nCategories nCells"		name_in_code="iceEnthalpyMask"/>
 		<var name="iceSalinityMask"			type="integer"	dimensions="nIceLayers nCategories nCells"		name_in_code="iceSalinityMask"/>
 		<var name="snowEnthalpyMask"			type="integer"	dimensions="nSnowLayers nCategories nCells"		name_in_code="snowEnthalpyMask"/>
+		<var name="snowIceMassMask"			type="integer"	dimensions="nSnowLayers nCategories nCells"		name_in_code="snowIceMassMask"			packages="pkgColumnTracerSnowDensity"/>
+		<var name="snowLiquidMassMask"			type="integer"	dimensions="nSnowLayers nCategories nCells"		name_in_code="snowLiquidMassMask"		packages="pkgColumnTracerSnowDensity"/>
+		<var name="snowGrainRadiusMask"		type="integer"	dimensions="nSnowLayers nCategories nCells"		name_in_code="snowGrainRadiusMask"		packages="pkgColumnTracerSnowGrainRadius"/>
```
Comment (akturner) [inline src/core_seaice/Registry.xml]: Fix indent

File: src/core_seaice/Registry.xml [hunk 13]
```diff
@@ -3635,6 +3821,20 @@
...
+	<!-- snow -->
+	<var_struct name="snow" time_levs="1" packages="pkgColumnPackage">
+		<var name="snowDensityViaContent"		type="real"	dimensions="nCells Time"			name_in_code="snowDensityViaContent"		units="kg/m^3"/>
+		<var name="snowDensityViaCompaction"		type="real"	dimensions="nCells Time"			name_in_code="snowDensityViaCompaction"	units="kg/m^3"/>
+		<var name="snowLossToLeads"			type="real"	dimensions="nCells Time"			name_in_code="snowLossToLeads"			units="kg/m^2/s"/>
+		<var name="snowMeltMassCell"			type="real"	dimensions="nCells Time"			name_in_code="snowMeltMassCell"		units="kg/m^2/d"/>
+		<var name="snowMeltMassCategory"		type="real"	dimensions="nCategories nCells Time"		name_in_code="snowMeltMassCategory"		units="kg/m^2"/>
+		<var name="snowRadiusRadiationCategory"	type="real"	dimensions="nCategories nCells Time"		name_in_code="snowRadiusRadiationCategory"	units="um"/>
+		<var name="snowRadiusRadiation"		type="real"	dimensions="nCells Time"			name_in_code="snowRadiusRadiation"		units="um"/>
+		<var name="tau"				type="real"	dimensions="sno_dns dTdz T"			name_in_code="snowPropertyTau"		units="um"/>
```
Comment (akturner) [inline src/core_seaice/Registry.xml]: rename

File: src/core_seaice/Registry.xml [hunk 14]
```diff
@@ -3635,6 +3821,20 @@
...
+		<var name="kappa"				type="real"	dimensions="sno_dns dTdz T"			name_in_code="snowPropertyKappa"	units="none"/>
```
Comment (akturner) [inline src/core_seaice/Registry.xml]: rename

File: src/core_seaice/Registry.xml [hunk 15]
```diff
@@ -3635,6 +3821,20 @@
...
+		<var name="drdsdt0"				type="real"	dimensions="sno_dns dTdz T"			name_in_code="snowPropertyRate"	units="um/hr"/>
```
Comment (akturner) [inline src/core_seaice/Registry.xml]: rename

File: src/core_seaice/column/constants/cice/ice_constants_colpkg.F90 [hunk 16]
```diff
@@ -1,4 +1,4 @@
-!  SVN:$Id: ice_constants_colpkg.F90 1012 2015-06-26 12:34:09Z eclare $
+!  SVN:$Id: ice_constants_colpkg.F90 1030 2015-07-27 15:05:32Z eclare $
```
Comment (akturner) [inline src/core_seaice/column/constants/cice/ice_constants_colpkg.F90]: Does this change signify something?

File: src/core_seaice/column/ice_atmo.F90 [hunk 17]
```diff
@@ -1,4 +1,4 @@
-!  SVN:$Id: ice_atmo.F90 1182 2017-03-16 19:29:26Z njeffery $
+!  SVN:$Id: ice_atmo.F90 1196 2017-04-18 13:32:23Z eclare $
```
Comment (akturner) [inline src/core_seaice/column/ice_atmo.F90]: This (along with others like it) seems unnecessary?

File: src/core_seaice/column/ice_snow.F90 [hunk 18]
```diff
@@ -0,0 +1,902 @@
...
+      tmp5 = tmp5 + esnw_susp*alost
+
+      !-----------------------------------------------------------------
+      ! energy of suspended snow lost in leads
+      !-----------------------------------------------------------------
+      elost = tmp3 - tmp4
+      fhocn = fhocn + elost / dt
+
+      if (abs(tmp5) > nslyr*Lfresh*puny) then
+         print*,'energy conservation error in snow_redist', tmp3, tmp4, tmp5
```
Comment (akturner) [inline src/core_seaice/column/ice_snow.F90]: Is this going to work with MPAS error messaging?

File: src/core_seaice/column/ice_snow.F90 [hunk 19]
```diff
@@ -0,0 +1,902 @@
...
+            tmp2 = tmp2 + vsn(n)
+         enddo
+      endif
+
+      if (tmp2 < tmp1) fresh = fresh + rhos*(tmp1-tmp2)/dt
+
+      tmp2 = tmp2 + (mlost/rhos)
+
+      if (abs(tmp1-tmp2) > puny) then
+         print*,'mass conservation error in snow_redist', tmp1, tmp2
```
Comment (akturner) [inline src/core_seaice/column/ice_snow.F90]: Is this going to work this the E3SM messaging system?

File: src/core_seaice/model_forward/mpas_seaice_core_interface.F [hunk 20]
```diff
@@ -442,6 +452,9 @@ subroutine setup_packages_column_physics(configPool, packagePool, ierr)!{{{
      pkgTracerZAerosolsActive            = config_use_zaerosols
      pkgTracerZSalinityActive            = config_use_vertical_zsalinity
 
+     pkgColumnTracerSnowDensityActive    = config_use_snow_density
+     pkgColumnTracerSnowGrainRadiusActive= config_use_snow_grain_radius
```
Comment (akturner) [inline src/core_seaice/model_forward/mpas_seaice_core_interface.F]: Reformat with correct spacing

File: src/core_seaice/model_forward/mpas_seaice_core_interface.F [hunk 21]
```diff
@@ -473,6 +486,8 @@ subroutine setup_packages_column_physics(configPool, packagePool, ierr)!{{{
         pkgTracerVerticalIronActive         = .false.
         pkgTracerZAerosolsActive            = .false.
         pkgTracerZSalinityActive            = .false.
+        pkgColumnTracerSnowDensityActive    = .false.
+        pkgColumnTracerSnowGrainRadiusActive= .false.
```
Comment (akturner) [inline src/core_seaice/model_forward/mpas_seaice_core_interface.F]: Reformat with correct formatting

File: src/core_seaice/shared/mpas_seaice_column.F [hunk 22]
```diff
@@ -421,6 +431,170 @@ subroutine init_column_thermodynamic_profiles(domain)
 
   end subroutine init_column_thermodynamic_profiles
 
+!|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
+!
+!  init_column_snow
+!
+!> \brief
```
Comment (akturner) [inline src/core_seaice/shared/mpas_seaice_column.F]: Add brief and long description

File: src/core_seaice/shared/mpas_seaice_column.F [hunk 23]
```diff
@@ -421,6 +431,170 @@ subroutine init_column_thermodynamic_profiles(domain)
...
+!> \author Nicole Jeffery, LANL
+!> \date 1 April 2017
+!> \details
+!>
+!
+!-----------------------------------------------------------------------
+
+  subroutine init_column_snow(domain)
```
Comment (akturner) [inline src/core_seaice/shared/mpas_seaice_column.F]: Ambiguous name, since snow already included in the model?

File: src/core_seaice/shared/mpas_seaice_column.F [hunk 24]
```diff
@@ -2088,6 +2301,250 @@ subroutine column_prep_radiation(domain)
 
   end subroutine column_prep_radiation
 
+!|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
+!
+!  column_snow
+!
+!> \brief
```
Comment (akturner) [inline src/core_seaice/shared/mpas_seaice_column.F]: Add brief and details

File: src/core_seaice/shared/mpas_seaice_column.F [hunk 25]
```diff
@@ -2136,8 +2594,11 @@ subroutine column_radiation(domain, clock, lInitialization)
          config_use_shortwave_bioabsorption, &
          config_use_brine, &
          config_use_modal_aerosols, &
-         config_use_zaerosols, &
-         config_use_column_biogeochemistry
+         config_use_column_biogeochemistry!, &
+!         config_use_snow_redistribution
```
Comment (akturner) [inline src/core_seaice/shared/mpas_seaice_column.F]: Remove unused comment

File: testing_and_setup/seaice/configurations/snowMods/README [hunk 26]
```diff
@@ -0,0 +1,2 @@
+Standard physics configuration with full snow tracers and modifications.
```
Comment (akturner) [inline testing_and_setup/seaice/configurations/snowMods/README]: change snowMods to a more meaningful name and use underscore not camelcase

File: testing_and_setup/seaice/testing/testsuites/testsuite.snowMods.xml [hunk 27]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (akturner) [inline testing_and_setup/seaice/testing/testsuites/testsuite.snowMods.xml]: change snowMods as above

File: src/core_seaice/column/ice_itd.F90 [hunk 28]
```diff
@@ -1,4 +1,4 @@
-!  SVN:$Id: ice_itd.F90 1182 2017-03-16 19:29:26Z njeffery $
+!  SVN:$Id: ice_itd.F90 1196 2017-04-18 13:32:23Z eclare $
```
Comment (akturner) [inline src/core_seaice/column/ice_itd.F90]: Why are these changes here????????????

File: src/core_seaice/column/ice_mechred.F90 [hunk 29]
```diff
@@ -1,4 +1,4 @@
-!  SVN:$Id: ice_mechred.F90 1182 2017-03-16 19:29:26Z njeffery $
+!  SVN:$Id: ice_mechred.F90 1196 2017-04-18 13:32:23Z eclare $
```
Comment (akturner) [inline src/core_seaice/column/ice_mechred.F90]: Why?????

File: src/core_seaice/model_forward/mpas_seaice_core_interface.F [hunk 30]
```diff
@@ -442,6 +452,9 @@ subroutine setup_packages_column_physics(configPool, packagePool, ierr)!{{{
      pkgTracerZAerosolsActive            = config_use_zaerosols
      pkgTracerZSalinityActive            = config_use_vertical_zsalinity
 
+     pkgColumnTracerSnowDensityActive     = config_use_effective_snow_density
```
Comment (akturner) [inline src/core_seaice/model_forward/mpas_seaice_core_interface.F]: Change pkgColumnTracerSnowDensityActive to pkgColumnTracerEffectiveSnowDensityActive

File: src/core_seaice/shared/mpas_seaice_column.F [hunk 31]
```diff
@@ -421,6 +431,178 @@ subroutine init_column_thermodynamic_profiles(domain)
 
   end subroutine init_column_thermodynamic_profiles
 
+!|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
+!
+!  init_column_snow_tracers
+!
+!> \brief
+!> Initializes snow physics tracers
```
Comment (akturner) [inline src/core_seaice/shared/mpas_seaice_column.F]: Brief should be on same line as the \brief tag

File: src/core_seaice/shared/mpas_seaice_column.F [hunk 32]
```diff
@@ -2088,6 +2309,261 @@ subroutine column_prep_radiation(domain)
 
   end subroutine column_prep_radiation
 
+!|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
+!
+!  column_snow
+!
+!> \brief
+!>
+!> Enable snow grain aging, effective snow density, wind compaction and redistribution
```
Comment (akturner) [inline src/core_seaice/shared/mpas_seaice_column.F]: Brief should be on same line as \brief tag

File: src/core_seaice/Registry.xml [hunk 33]
```diff
@@ -2232,6 +2326,10 @@
 		<var name="iceEnthalpy"			type="real"	dimensions="nIceLayers nCategories nCells Time"		name_in_code="iceEnthalpy"/>
 		<var name="iceSalinity"			type="real"	dimensions="nIceLayers nCategories nCells Time"		name_in_code="iceSalinity"/>
 		<var name="snowEnthalpy"			type="real"	dimensions="nSnowLayers nCategories nCells Time"		name_in_code="snowEnthalpy"/>
+		<var name="snowIceMass"			type="real"	dimensions="nSnowLayers nCategories nCells Time"	units="kg/m^3"	description="Mass of ice in snow layer"	name_in_code="snowIceMass"	packages="pkgColumnTracerSnowDensity"/>
```
Comment (akturner) [inline src/core_seaice/Registry.xml]: Descriptions for new vars (except advection ones!)

**Comment by njeffery** [comment]
@akturner:  finished corrections.  The SnowAgingPropertiesInput  file is  /turquoise/usr/projects/climate/njeffery/DATA/snow/snicar_drdt_bst_fit_60_c04252019.nc
